### PR TITLE
Windows: Allow `USER` in builder

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

@thaJeztah @vieux Another one which should strongly be considered for 1.13.

Fixes https://github.com/docker/docker/issues/28027. `USER` is supported in Windows now after https://github.com/docker/docker/pull/28184, but the block in the builder was never removed.

This removes the block and adds a CI test for it.

@duglin FYI

### CODE REVIEW

**Comment by jhowardmsft** [comment]
@johnstep

**Comment by duglin** [comment]
@jhowardmsft are we able to enable any of the existing builder/USER tests for windows now?

**Comment by jhowardmsft** [comment]
@duglin - perhaps, but a lot of rework.

**Review by justincormack — APPROVED** [review]

**Comment by duglin** [comment]
wouldn't mind seeing some more tests, like setting it multiple times to make sure it sticks - during the build and then at run time.  But LGTM for now.

**Comment by duglin** [comment]
I should say, LGTM once janky finishes

**Comment by thaJeztah** [comment]
I'm ok with adding this, its better to do it now than later, but @vieux @duglin LMK if you disagree

**Comment by jhowardmsft** [comment]
@duglin 

> wouldn't mind seeing some more tests, like setting it multiple times to make sure it sticks

Reminds me of what @tianon said in a maintainers meeting a few weeks back - "If you give a moose a muffin, he's going to want a glass of milk" 😸

**Comment by duglin** [comment]
ouch! :-)  you gotta admit, that one test is pretty slim.  That's why I was asking about enabling the other ones.

**Comment by tianon** [comment]
[image](https://www.amazon.com/If-You-Give-Moose-Muffin/dp/0060244054)

**Comment by jhowardmsft** [comment]
@duglin - yes, myself and Darren are slowly working through enabling more and more tests on Windows - it will be a few more months before we are there. I just added a basic test for 1.13 to ensure that the functionality does at least minimally work.

**Comment by soanni** [comment]
team, after merging can i try it in nightly build? this PR will save my bacon

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by darrenstahlmsft** [comment]
@soanni You can download the latest master builds at [master.dockerproject.org](master.dockerproject.org). This change will be available there very shortly after it is merged.

**Comment by thaJeztah** [comment]
@jhowardmsft @darrenstahlmsft because we don't have the tests running, so not able to verify; how likely is it that the behavior will change (possibly breaking)?

**Comment by jhowardmsft** [comment]
@thaJeztah Not sure what you're asking - there is a test in there.... Chances of changing are minimal to zero. The API is fixed on the HCS side and part of Windows.

**Comment by soanni** [comment]
sorry to bother you team. do you have any dates when this PR will be merged?)

**Comment by jhowardmsft** [comment]
@vieux @tiborvass ^^

**Comment by LK4D4** [comment]
LGTM

**Comment by soanni** [comment]
it seems like the build for 1.13 wasn't placed on https://master.dockerproject.org/. The last build date is November 12th

**Comment by thaJeztah** [comment]
@soanni https://master.dockerproject.org/ is built from "master"; now that the 1.13 release branch is created, the version on master is bumped to 1.14-dev, so you need to look for 1.14-dev builds now.

We should remove the old 1.13-dev builds from https://master.dockerproject.org/, I'll open an issue for that.

**Comment by thaJeztah** [comment]
Opened https://github.com/docker/docker/issues/28579 for removing the old builds

**Comment by soanni** [comment]
thanks @thaJeztah 

if i'm adding user without password i'm able to switch user
------------------
FROM microsoft/windowsservercore:latest

# add sp_farm user, add him to Administrators group
RUN NET USER sp_farm /add
RUN	NET LOCALGROUP Administrators /add sp_farm

USER sp_farm
SHELL ["powershell"]
RUN whoami
------------------------

but if i'm adding user with password i got the error. Or i need to start-process powershell and provide PSCredentials object. It's not clear how i can do it in Dockerfile during image build.

FROM microsoft/windowsservercore:latest

# add sp_farm user, add him to Administrators group
RUN NET USER sp_farm "Password1234" /add
RUN	NET LOCALGROUP Administrators /add sp_farm

USER sp_farm
SHELL ["powershell"]
RUN whoami

image

**Comment by thaJeztah** [comment]
Thanks @soanni, could you open an issue for that? It's easier for us to have a separate issue that we can assign / prioritize

**Comment by soanni** [comment]
sure

**Comment by soanni** [comment]
could please anyone tell me what's the planned release date for 1.14 where USER instruction enabled? thanks in advance

**Pull request merged**
